### ISSUE TITLE
ReactiveObject analogy

### ISSUE THREAD

**Issue by TomKaltz** [WebRxJS/WebRx#41]
I'm glad I happened upon this framework.  I always loved ReactiveUI in C#. So happy it exists in JS.

I'm wondering if there is a semantic where I can watch for property changes at the viewmodel level just like ReactiveObject in ReactiveUI.

**Comment by oliverw** [comment]
Sure, it's called <code>wx.observeObject</code>. Unfortunately the function is not yet exposed in web.rx.d.ts but you can check out the signature here: https://github.com/WebRxJS/WebRx/blob/master/src/Core/Utils.ts#L509

**Comment by TomKaltz** [comment]
observeObject does not observe the ``wx.list()`` in an object, correct?

**Comment by oliverw** [comment]
That's true but if what you really what to achieve is to track changes to all objects contained in an observable list then all you have to do is set <code>myList.changeTrackingEnabled = true</code> and subscribe to the lists <code>itemChanged</code> observable.

**Comment by TomKaltz** [comment]
I am looking to watch a model object that contains wx properties and lists
for changes so that I can synchronize with local storage on demand.
Regarding this, should standard serialization work with these types?  I am
using plain JS btw.

On Monday, April 11, 2016, Oliver Weichhold <notifications@github.com>
wrote:

> That's true but if what you really what to achieve is to track changes to
> all objects contained in an observable list then all you have to do is set myList.changeTrackingEnabled
> = true and subscribe to the lists itemChanged observable.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/WebRxJS/WebRx/issues/41#issuecomment-208375399>
>


-- 
Thomas Kaltz III
586-214-7150
Sent from Gmail Mobile

**Comment by oliverw** [comment]
You should be able to get the desired result by observing changes to serialization-relevant properties using <code>whenAny</code> (see http://webrxjs.org/docs/output-properties.html) and [merging](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/merge.md) the observable returned from whenAny with the list's <code>itemChanged</code> observable.

I'm afraid standard serialization will not work since wx.property creates a function member instead of a field. But this should work:

```javascript
function ViewModel2Object(vm) {
	var result = {};

	// non observable fields
	Object.keys(vm).forEach(function(x)) {
		let prop = vm[x];
		if(!wx.queryInterface(prop, wx.IID.IObservableProperty)) {
			result[x] = prop;
		}
	});

	// now add observable property values
	var observableProperties = wx.getOwnPropertiesImplementingInterface(vm, wx.IID.IObservableProperty);
	observableProperties.forEach(function(x) {
		result[x.propertyName] = x.property();
	});
}
```
